spark3/src/main/java/org/apache/iceberg/spark/procedures/SnapshotProcedure.java

```diff
@@ -0,0 +1,114 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.procedures;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.iceberg.actions.SnapshotAction;
+import org.apache.iceberg.actions.Spark3SnapshotAction;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.spark.Spark3Util.CatalogAndIdentifier;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.iceberg.catalog.ProcedureParameter;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import scala.runtime.BoxedUnit;
+
+class SnapshotProcedure extends BaseProcedure {
+  private static final DataType MAP = DataTypes.createMapType(DataTypes.StringType, DataTypes.StringType);
+  private static final ProcedureParameter[] PARAMETERS = new ProcedureParameter[]{
+      ProcedureParameter.required("snapshot_source", DataTypes.StringType),
+      ProcedureParameter.required("table", DataTypes.StringType),
+      ProcedureParameter.optional("table_location", DataTypes.StringType),
+      ProcedureParameter.optional("table_options", MAP)
+  };
+
+  private static final StructType OUTPUT_TYPE = new StructType(new StructField[]{
+      new StructField("num_datafiles_included", DataTypes.LongType, false, Metadata.empty())
+  });
+
+  private SnapshotProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  public static SparkProcedures.ProcedureBuilder builder() {
+    return new BaseProcedure.Builder<SnapshotProcedure>() {
+      @Override
+      protected SnapshotProcedure doBuild() {
+        return new SnapshotProcedure(tableCatalog());
+      }
+    };
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    String source = args.getString(0);
+    String dest = args.getString(1);
+
+    String snapshotLocation = args.isNullAt(2) ? null : args.getString(2);
+
+    Map<String, String> options = new HashMap<>();
+    if (!args.isNullAt(3)) {
+      args.getMap(3).foreach(DataTypes.StringType, DataTypes.StringType,
```

Comment:
Calling scala from java

Reply:
Yeah, maybe we should rely on Scala conversions and then do all the logic in Java? I'm not sure what impact this may have on compatibility, but I'd rather not break because Scala changes something.

Reply:
This is a little difficult because it's the Scala method on InternalRow which doesn't really return a map, it returns "catalyst.util.MapData", so it's not a real map and the method we are using is

```scala
  def foreach(keyType: DataType, valueType: DataType, f: (Any, Any) => Unit): Unit = {
```

I thought about writing another method to convert MapData to Java maps but that also seems brittle. I think the safest thing to do here is use the Scala Method for MapData.foreach

Reply:
Sounds reasonable to me.